torch/csrc/jit/codegen/cuda/fusion_executor.cpp

```diff
@@ -0,0 +1,589 @@
+      "Wrong number of kernel outputs.");
+  for (size_t i = 0; i < outputs.size(); ++i) {
+    const at::Tensor& arg = outputs[i];
+    const Val* param = fusion->outputs()[i];
+    std::stringstream msg;
+    TORCH_INTERNAL_ASSERT(
+        validateKernelArgTensor(arg, param, device, msg),
+        "Output argument at position ",
+        i,
+        " is invalid; ",
+        msg.str());
+  }
+}
+
+}; // namespace
+
+int FusionExecutor::fusion_id_counter = 0;
+
+std::string FusionExecutor::getKernel() {
+  // generating cuda code;
+  std::string code = std::string("namespace ") + FusionExecutor::Namespace() +
+      " {\n" + KernelUtils() + GPULower(&fusion_).getKernel(KernelName()) +
+      "}\n";
+
+  const char* debug_env = getenv("PYTORCH_CUDA_FUSER_DEBUG");
+  if (debug_env && atoi(debug_env)) {
+    std::cout << "\n==== codegen output for kernel: " << KernelName()
+              << " ====" << std::endl
+              << code << std::endl
+              << "====================================" << std::endl;
+  }
+
+  return code;
+}
+
+// TODO: add options for constraints, if we want to enforce a parallelization
+// strategy because we did something like split on a symbolic size.
+LaunchParams FusionExecutor::computeLaunchParams(
+    const at::ArrayRef<IValue>& aten_inputs) {
+  TORCH_INTERNAL_ASSERT(
+      fusion_.inputs().size() == aten_inputs.size(),
+      "Something went wrong configuring launch. Inputs no longer match.");
+
+  auto fusion_inputs = fusion_.inputs();
+  EvaluationContext ec(&fusion_);
+
+  // This should probably move to EvaluationContext as we may want to bind input
+  // values frequently. Bind fusion input values to runtime values.
+  for (size_t i = 0; i < fusion_.inputs().size(); i++) {
+    if (fusion_.inputs()[i]->getValType() == ValType::TensorView) {
+      TensorView* cg_tensor = fusion_.inputs()[i]->as<TensorView>();
+
+      TORCH_INTERNAL_ASSERT(
+          aten_inputs[i].isTensor(),
+          "Something went wrong configuring launch. Inputs no longer match.");
+
+      auto aten_tensor = aten_inputs[i].toTensor();
+
+      TORCH_INTERNAL_ASSERT(
+          aten_tensor.ndimension() == cg_tensor->getRootDomain().size(),
+          "Something went wrong configuring launch. Inputs no longer match.");
+
+      for (size_t dim = 0; dim < cg_tensor->getRootDomain().size(); dim++) {
+        ec.bind(
+            cg_tensor->getRootDomain()[dim]->extent(),
+            aten_tensor.sizes()[dim]);
+      }
+    }
+  }
+
+  LaunchParams lp;
+
+  // Grab all used values and run through them
+  auto unordered_vals = DependencyCheck::getAllValsBetween(
+      {fusion_inputs.begin(), fusion_inputs.end()}, fusion_.outputs());
+
+  // Grab only the TensorViews
+  std::set<TensorView*> unordered_tvs;
+  for (auto val : unordered_vals) {
+    if (val->getValType().value() == ValType::TensorView) {
+      TensorView* tv = val->as<TensorView>();
+      unordered_tvs.emplace(tv);
+    }
+  }
+
+  // Infer bindings
```

Comment:
I like it that we are inferring binding/launch_config from the codegen IR.
My question here is, would the same thing work for things like: dynamic shared memory size & global buffer size?

Reply:
Yes, this is precisely how we will do it. I think everything will go together so it can use the same expression context.